### ISSUE TITLE
Servant Death by NP (Arash, Chen Gong, etc)

### ISSUE THREAD

**Issue by the3dsandwich** [Fate-Grand-Automata/FGA#736]
Before opening an issue:
- &#x2612; No one has asked for this feature before in other issues (even the closed ones)

> If you open unnecessary/dumb issues, you'll be rewarded with appropriate tags.

I know this topic is discussed in many other issues (all closed now and I probably shouldn't be replying there) but I think I have a suggestion that might work and would like to ask for your opinion.

Suggested approach: **Detect servant death by checking when there are exactly 2/3 sets of the servant HP/NP bar**, and start tapping.

(I know there is a list of the existing issues that asked about the same problem)
#438 #437 #393 #345 #185

So far on those issues we see suggest the following approaches:

- Wait for X seconds then click for 300ms, X is configurable (#345): doesn't work because servants don't die every turn and there needs to be a way to detect servant death
- Add "death" option in autoskill, then script knows when to expect a servant death (#437): doesn't work because we still don't know how long the NP lasts and we also don't know how to detect a servant death.

I think I have a way to detect a servant death, or rather, the timing when we should start tapping.

https://www.youtube.com/watch?v=jzxE6adbgxg

Here's a run of using Chen Gong where each wave a servant dies to illustrate. 

- Let's say there are 3 servants on the field before Chen Gong's NP and there is at least one servant behind.
- After Chen Gong's NP, instead of trying to detect the servant sprite dying, we can instead match/look for the presence of the HP/NP bar (or the LV text underneath). If we see exactly 2 occurrences on the screen, then we know one of them was used as sacrifice and the animation has not finished yet: **we start our tap here**
- We also know that when the animation finishes, there will be again 3 occurrences of the HP/NP bar, plus there will be the attack button.
- If there are not enough servants to replace, then it won't be a problem either since when we check the occurrences of the HP/NP bars we also check whether the attack button is present.

I meant the red circle when I say the HP/NP bar.

![Untitled](https://user-images.githubusercontent.com/32169552/122664818-4245a880-d1d6-11eb-89ec-c42da38686b3.png)

So I suggest the logic be a combination of the previous suggestions:

1. Add "death" option in autoskill so FGA knows when it should anticipate and detect servant deaths.
2. After an NP, if the "death" option is enabled, FGA waits for a configurable X seconds before matching images on the screen.
3. FGA checks 4 regions: servant 1's HP/NP bar (a), servant 2's HP/NP bar (b), servant 3's HP/NP bar (c), the attack button (d).
4. if d matches, next turn/wave. No taps.
5. if the number of matches for a,b,c is between 1 and 2, initiate taps.

![Mikon Anticipate Her Sacrifice While Master Test Out FGA - YouTube and 6 more pages - Personal - Microsoft​ Edge 2021-06-20 14-08-06_1](https://user-images.githubusercontent.com/32169552/122664381-a2871b00-d1d3-11eb-91da-fd886bc3f41a.gif)

I wonder if we are able to switch to a different image matching entry point when it's controlled by autoskill. Maybe we might even skip the autoskill check and instead match and count occurrences directly. I cannot be sure. But if this works then it'll be so nice. Thanks again!!!!

**Comment by MathewSachin** [comment]
> I wonder if we are able to switch to a different image matching entry point when it's controlled by autoskill.

AutoSkill never controls the matching. There's an outer loop that runs AutoSkill if it detects the Attack button on the battle screen.
So, AutoSkill has to exit after clicking on cards and NPs to let the outer loop run again.
This is needed to handle different cases that can happen after NPs/cards are clicked:
1. **Next turn/battle:** Outer loop detects the Attack button, runs AutoSkill.
2. **Battle ends:** Outer loop detects result screen, taps through the result screens.
3. **Connection error:** This can happen in cases like using NP of a new servant for the first time. Clicks on retry.

> 4. if d matches, next turn/wave. No taps.

Due to the above reasons, we can't just match (d) or make it an AutoSkill feature without duplicated code.

> 5. if the number of matches for a,b,c is between 1 and 2, initiate taps.

There could be cases where only 2 servants were left already. In that case, we would have to integrate this with ServantTracker.kt to detect how many servants are present on the field. Even then, ServantTracker.kt assumes that you always have 6 servants in your party, which doesn't matter in case of skill/NP spam, but there's a possibility of unnecessary clicks in this case.

---

Overall, I think this is still a very complex solution to pull off for very minimal time saved. IMHO, it is better to not run a Gong/Arash/plugsuit comp with FGA for people who're worried about the time it takes 😆.

**Comment by the3dsandwich** [comment]
What if we don't check for autoskill and instead add just an additional screen in AutoBattle.loop() where it checks (1) Attack button is not there and (2) less than 3 servants are on the field, and when that happened, click once instead of a bunch of clicks?

I realized this scenario should only appear during a dying scene where inputs are ignored anyways, so there shouldn't be a need to touch autoskill.

We also don't need to care about the situation where there are actually 2 servants (or 1) left, since the click will only occur during an animation.

---

So basically in AutoBattle.loop(), we add a new condition (no attack button and 2 servants alive) then make a single click. Will this be doable? Or did I miss a scenario?

**Comment by MathewSachin** [comment]
Sounds good to me!
Also, this new condition should be added into the loop after the Battle screen check, that way we already know that we're not on the Battle screen if this check is triggered and just need to check for the HP/NP icons.

**Comment by the3dsandwich** [comment]
Cool! I'll see if I can make some changes on my side these couple days. I'll test some of it and see if this approach actually works.

**Comment by MathewSachin** [comment]
I've pushed adjustments for JP. The center click on JP clicked on the middle servant and opened up the portrait, so I changed that location.
Also, I made the region to check for individual servants bigger because on JP, the UI shifts a bit vertically depending on whether the device is wider than 18:9.

**Comment by the3dsandwich** [comment]
Ah, makes sense. Thanks for everything on this matter, I know this has been brought up so many times and I'm just glad we found a way to handle this 😊

**Comment by MathewSachin** [comment]
Hey, I sent you an invitation to this GitHub org. You help us out a lot recently! Really appreciate that. 😄

**Comment by the3dsandwich** [comment]
Thank you so much! I'll see how I can help more. Really appreciate that!
